docs/layout.rst

```diff
@@ -0,0 +1,246 @@
+
+In Taichi, the layout is defined in a recursive manner. See :ref:`snode` for more details about how this works. We suggested starting with the default layout specification (simply by specifying ``shape`` when creating tensors using ``ti.var/Vector/Matrix``),
+and then migrate to more advanced layouts using the ``ti.root.X`` syntax.
+
+Taichi decouples algorithms from data layouts, and the Taichi compiler automatically optimizes data accesses on a specific data layout. These Taichi features allow programmers to quickly experiment with different data layouts and figure out the most efficient one on a specific task and computer architecture.
+
+
+From ``shape`` to ``ti.root.X``
+-------------------------------
+
+For example, this declares a 0-D tensor:
+
+.. code-block:: python
+
+    x = ti.var(ti.f32)
+    ti.root.place(x)
+    # is equivalent to:
+    x = ti.var(ti.f32, shape=())
+
+This declares a 1D tensor of size ``3``:
+
+.. code-block:: python
+
+    x = ti.var(ti.f32)
+    ti.root.dense(ti.i, 3).place(x)
+    # is equivalent to:
+    x = ti.var(ti.f32, shape=3)
+
+This declares a 2D tensor of shape ``(3, 4)``:
+
+.. code-block:: python
+
+    x = ti.var(ti.f32)
+    ti.root.dense(ti.ij, (3, 4)).place(x)
+    # is equivalent to:
+    x = ti.var(ti.f32, shape=(3, 4))
+
+You may wonder, why not simply specify the ``shape`` of the tensor? Why bother using the more complex version?
+Good question, let go forward and figure out why.
+
+
+Row-major versus column-major
+-----------------------------
+
+Let's start with the simplest layout.
+
+Since address spaces are linear in modern computers, for 1D Taichi tensors, the address of the ``i``-th element is simply ``i``.
+
+To store a multi-dimensional tensor, however, it has to be flattened, in order to fit into the 1D address space.
+For example, to store a 2D tensor of size ``(3, 2)``, there are two ways to do this:
+
+    1. The address of ``(i, j)``-th is ``base + i * 2 + j`` (row-major).
+
+    2. The address of ``(i, j)``-th is ``base + j * 3 + i`` (column-major).
+
+To specify which layout to use in Taichi:
+
+.. code-block:: python
+
+    ti.root.dense(ti.i, 3).dense(ti.j, 2).place(x)    # row-major (default)
+    ti.root.dense(ti.j, 2).dense(ti.i, 3).place(y)    # column-major
+
+Both ``x`` and ``y`` have the same shape of ``(3, 2)``, and they can be accessed in the same manner, where ``0 <= i < 3 && 0 <= j < 2``. They can be accessed in the same manner: ``x[i, j]`` and ``y[i, j]``.
+However, they have a very different memory layouts:
+
+.. code-block::
+
+    #     address low ........................... address high
+    # x:  x[0,0]   x[0,1]   x[0,2] | x[1,0]   x[1,1]   x[1,2]
+    # y:  y[0,0]   y[1,0] | y[0,1]   y[1,1] | y[0,2]   y[1,2]
+
+See? ``x`` first increases the first index, while ``y`` first increases the second index. When overflow, reset to zero and begin to increase the second index.
```

Comment:
WDYT if we write this like "`x` first increases along the `i` axis (or index), while `y` .." to be more explicit?